### PR TITLE
add xbfmg utility and corresponding tests

### LINKED ISSUE #111: consider adding xbfmg utility to NCEPLIBS-bufr utils
`xbfmg` is a WCOSS utility that breaks apart a file of BUFR messages into files of individual BUFR messages, so that they can be worked with individually rather than as one giant file.

re: [hpc-stack#176](https://github.com/NOAA-EMC/hpc-stack/issues/176), this utility may be useful enough to a wider audience that it makes sense to add it to the NCEPLIBS-bufr utils package, as was done previously with the `debufr` utility.

### PR DESCRIPTION
**Pull request by jbathegit**
Fixes #111

Part of #32 

I've also included two tests with this new utility.

### CODE REVIEW

**Comment by jbathegit** [comment]
Overall test coverage is now up to 66%, but xbfmg is C code, so it doesn't show up in the gcovr output since that only analyzes Fortran code at the moment.

**Comment by jbathegit** [comment]
Hello @rmclaren @kgerheiser @edwardhartnett @jack-woollen @aerorahul 

Could I trouble at least one of you to please take a look at this sometime in the next day or two and let me know if you have any questions or concerns?

Thanks!

File: utils/xbfmg.c [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+        /*
+	**  Get the valid options from the command line:
+	*/
+        while ( ( ch = getopt ( argc, argv, "vgh" ) ) != EOF ) {
+            switch ( ch ) {
+		case 'v':
+		    bvers ( bvstr, sizeof(bvstr) );
+		    /* append a trailing NULL to bvstr for printf */
+		    for ( ii = 0; ii < sizeof(bvstr); ii++ ) {
```
Comment (rmclaren) [inline utils/xbfmg.c:138]: why not declare the indexer localy.... something like 

`for (unsigned int strIdx=0; strIdx < VersionStrSize; strIdx++)`.

File: utils/xbfmg.c [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+	int ch;
+
+	FILE *fp;
+
+	f77int msglen, wkint;
+	f77int c24 = 24, c1 = 1;
+
+	unsigned long i, filesize, noutfile;
+
+	unsigned short ii;
```
Comment (rmclaren) [inline utils/xbfmg.c:128]: No need to declare all your local variables at the beginning of the function like this.

File: utils/xbfmg.c [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+
+	char *pc, *pmsg, *psb;
+
+	int save_GTSbull = 0;
+
+	char outfile[MXFLEN];
+	char outfile_temp[MXFLEN];
+	char wkstr[MXFLEN];
+
+	char bvstr[9] = "        ";
```
Comment (rmclaren) [inline utils/xbfmg.c:117]: Why not initialize the array with null values. One way you might do this is with calloc;

`char* bvstr = (char*) calloc(VersionStrSize, sizeof(char));`

which initializes the allocated heap memory with zeros.

Comment (rmclaren) [reply utils/xbfmg.c:117]: Why not give bvstr a more descriptive name... versionStr (or something).

Comment (rmclaren) [reply utils/xbfmg.c:117]: Move this allocation to where you are using the string.

File: utils/xbfmg.c [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+ * <b>Program history log:</b>
+ * - 2018-03-01  J. Ator  -- Original author
+ * - 2021-09-29  J. Ator  -- Use basename instead of pid in output filenames
+ */
+
+main( int argc, char *argv[] ) {
+
+	struct stat fileinfo;
+
+	char *pc, *pmsg, *psb;
```
Comment (rmclaren) [inline utils/xbfmg.c:109]: use more descriptive names

File: utils/xbfmg.c [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+		        if ( ( bvstr[ii] != '.' ) && ( !isdigit(bvstr[ii]) ) ) {
+		          bvstr[ii] = '\0';
+		          break;
+		        }
+		    }
+		    printf( "This is xbfmg v3.2.0, built with BUFRLIB v%s\n", bvstr );
+		    return 0;
+                case 'g':
+		    save_GTSbull = 1;
+                    break;
```
Comment (rmclaren) [inline utils/xbfmg.c:148]: inconsistent indentation

File: utils/xbfmg.c [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+        while ( ( ch = getopt ( argc, argv, "vgh" ) ) != EOF ) {
+            switch ( ch ) {
+		case 'v':
+		    bvers ( bvstr, sizeof(bvstr) );
+		    /* append a trailing NULL to bvstr for printf */
+		    for ( ii = 0; ii < sizeof(bvstr); ii++ ) {
+		        if ( ( bvstr[ii] != '.' ) && ( !isdigit(bvstr[ii]) ) ) {
+		          bvstr[ii] = '\0';
+		          break;
+		        }
```
Comment (rmclaren) [inline utils/xbfmg.c:142]: might be able to get rid of all this if you prefill bvstr with null.. (depends what bvers does).

File: utils/xbfmg.c [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+                    break;
+		case 'h':
+                    printf( "\nPROGRAM %s\n", argv[0] );
+                    printf( "\nABSTRACT: This program reads an input file containing one or more\n" );
+		    printf( "  BUFR messages as given by the first argument.  It then extracts each\n" );
+		    printf( "  each individual BUFR message into its own separate output file within\n" );
+		    printf( "  the current working directory.\n" );
+                    prtusage( argv[0] );
+                    return 0;
+                    break;
```
Comment (rmclaren) [inline utils/xbfmg.c:157]: Indentation

File: utils/xbfmg.c [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+	} 
+	pc[i] = '\0';
+	fclose( fp );
+
+	/*
+ 	**  Create an output file name template.
+        */
+	strcpy( wkstr, argv[optind] );
+	strcpy( outfile_temp, basename( wkstr ) );
+	strcat( outfile_temp, ".xbfmg.out" );
```
Comment (rmclaren) [inline utils/xbfmg.c:205]: Would this work?

```
char* outfile_base =  basename(arcv[optind])
strcat(outfile_base, ".xbfmg.out")
```

File: utils/xbfmg.c [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+
+	/*
+	**  Call wrdlen function to initialize BUFRLIB and determine machine endianness.
+	*/
+	wrdlen( );
```
Comment (rmclaren) [inline utils/xbfmg.c:210]: why not have a function on bufrlib called bufr_initialize or something

**Review by rmclaren — APPROVED** [review]
I have some comments, but since this is just a utility program I'm not going to worry about it to much, especially since this codes been around since 2018. In general I have the following comments about the C code (I don't write a lot of straight C code, so bare with me):
- No need to pre-allocate all the variables at the beginning of the function.
- Could use better names for things... abreviations like pc, psmsg, bvers etc obfuscate the intention of the code.
- Whole program in the "main" function...

See code for other comments.

File: utils/xbfmg.c [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+
+	/*
+	**  Locate each BUFR message within the input file and write each one to a separate output file.
+	**
+	**  Note that we can't use the intrinsic C strstr function to locate the "BUFR" and "7777"
+	**  strings within the file, because the file could contain embedded NULL characters.
+	*/
+	noutfile = 0;
+	pmsg = psb = pc;
```
Comment (rmclaren) [inline utils/xbfmg.c:219]: You are doing `pmsg - pc` in many places. It might make the code clearer if you stored an offset instead of a pointer.

**Comment by jbathegit** [comment]
Thanks for your thoughtful and careful review @rmclaren!  I'll go ahead and merge this for now, but I like some of your suggestions and will look at incorporating them as part of a future PR.

File: utils/xbfmg.c (hunk 8 above)
Comment (jbathegit) [reply utils/xbfmg.c:205]: The basename() function can modify its input string, and I didn't want to modify argv[optind], which is why I wrote it the way I did.

File: utils/xbfmg.c (hunk 7 above)
Comment (jbathegit) [reply utils/xbfmg.c:157]: Will do in next PR

File: utils/xbfmg.c (hunk 5 above)
Comment (jbathegit) [reply utils/xbfmg.c:148]: Will do in next PR

File: utils/xbfmg.c (hunk 6 above)
Comment (jbathegit) [reply utils/xbfmg.c:142]: That's a very good suggestion - thanks!

Will do in next PR

File: utils/xbfmg.c (hunk 3 above)
Comment (jbathegit) [reply utils/xbfmg.c:117]: Will do in next PR

File: utils/xbfmg.c (hunk 1 above)
Comment (jbathegit) [reply utils/xbfmg.c:138]: Will do in next PR

**Pull request merged**